projects/sw-offline-google-analytics/src/index.js

```diff
@@ -0,0 +1,119 @@
+/*
+ Copyright 2016 Google Inc. All Rights Reserved.
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+ http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+/* eslint-env worker, serviceworker */
+
+const constants = require('./lib/constants.js');
+const enqueueRequest = require('./lib/enqueue-request.js');
+const log = require('../../../lib/log.js');
+const replayQueuedRequests = require('./lib/replay-queued-requests.js');
+
+/**
+ * In order to use the library, call`goog.offlineGoogleAnalytics.initialize()`.
+ * It will take care of setting up service worker `fetch` handlers to ensure
+ * that the Google Analytics JavaScript is available offline, and that any
+ * Google Analytics requests made while offline are saved (using `IndexedDB`)
+ * and retried the next time the service worker starts up.
+ *
+ * @example
+ * // This code should live inside your service worker JavaScript, ideally
+ * // before any other 'fetch' event handlers are defined:
+ *
+ * // First, import the library into the service worker global scope:
+ * importScripts('path/to/offline-google-analytics-import.js');
+ *
+ * // Then, call goog.offlineGoogleAnalytics.initialize():
+ * goog.offlineGoogleAnalytics.initialize({
+ *   parameterOverrides: {
+ *     // Optionally, pass in an Object with additional parameters that will be
+ *     // included in each replayed request.
+ *     // See https://developers.google.com/analytics/devguides/collection/protocol/v1/parameters
+ *     cd1: 'Some Value',
+ *     cd2: 'Some Other Value'
+ *   },
+ *   hitFilter: (searchParams) =>
+ *     // Sets the `qt` param as a custom metric.
+ *     const qt = searchParams.get('qt');
+ *     searchParams.set('qt', qt);
```

Comment:
Sorry, this is supposed to be `searchParams.set('cm1', qt);` (as it is in the test file).

This is a real example. It's being used here to track (via a custom metric) what the `qt` param was. This could be used in conjunction with an offline/online change handler to get a sense for how long (on average) users were without connectivity.

Another good example would be if a site had more than one tracker, and the dimension index they wanted to use to track service worker replays were different from each other. In that case they'd have to run logic like this:

```js
switch (searchParams.get('tid')) {
  case 'UA-XXXXX-Y':
    return 1; // or whatever this happens to be for property UA-XXXXX-Y
  case 'UA-XXXXX-Z'
    return 2; // or whatever this happens to be for property UA-XXXXX-Z
  ...
}
```